### PR TITLE
Adds configurable compression algorithms for chunks

### PR DESCRIPTION
**Pull request by cyriltovena**
This introduces a new chunk format (v2) that contains the current compression algorithm used to encode the chunk.

Loki can still read (v1) format which are by default `gzip`, and the new format is now the default with still `gzip` as encoding.

You can mix multiple encodings over time now, if you want to try different format.

We support those encoding now:
- gzip: default gzip golang implementation the slowest with the best compression ratio.
- gzip-1:  gzip with best speed configuration, a bit fast but also a lower compression ratio.
- lz4 : fast compression and decompression speed. (use more memory than snappy)
- none: no compressoion
- snappy: the popular google snappy compression format
- snappyv2: an optimized version of snappy see https://github.com/klauspost/compress/tree/master/s2

## Chunk size for each

```
  --- PASS: TestChunkSize/none (0.01s)
        Chunk size 2.7 MB
       characters  2623262
    --- PASS: TestChunkSize/gzip (0.01s)
        Chunk size 144 kB
        characters  2623262
    --- PASS: TestChunkSize/gzip-1 (0.01s)
         Chunk size 158 kB
        characters  2623262
    --- PASS: TestChunkSize/lz4 (0.00s)
        Chunk size 188 kB
        characters  2623262
    --- PASS: TestChunkSize/snappy (0.00s)
        Chunk size 272 kB
       characters  2623262
    --- PASS: TestChunkSize/snappyv2 (0.00s)
        Chunk size 205 kB
        characters  2623262
```

## Decompression speed benchmark

```
BenchmarkRead/none-16         	      15	  71535984 ns/op	134484800 B/op	  491005 allocs/op
--- BENCH: BenchmarkRead/none-16 bytes per second  1.8 GB
BenchmarkRead/gzip-16         	       5	 212703812 ns/op	134582374 B/op	  491465 allocs/op
--- BENCH: BenchmarkRead/gzip-16 bytes per second  617 MB
BenchmarkRead/gzip-1-16       	       5	 221152209 ns/op	134588478 B/op	  491452 allocs/op
--- BENCH: BenchmarkRead/gzip-1-16 bytes per second  593 MB
BenchmarkRead/lz4-16          	      10	 100376532 ns/op	141235551 B/op	  491049 allocs/op
--- BENCH: BenchmarkRead/lz4-16  bytes per second  1.3 GB
BenchmarkRead/snappy-16       	      10	 101250653 ns/op	134560112 B/op	  491038 allocs/op
--- BENCH: BenchmarkRead/snappy-16 bytes per second  1.3 GB
BenchmarkRead/snappyv2-16     	      10	 108484452 ns/op	137413780 B/op	  491057 allocs/op
--- BENCH: BenchmarkRead/snappyv2-16 bytes per second  1.2 GB
PASS
ok  	github.com/grafana/loki/pkg/chunkenc	31.598s
```

## Compression speed benchmark

```
pkg: github.com/grafana/loki/pkg/chunkenc
BenchmarkWrite/none-16         	     754	   2003727 ns/op	 2758437 B/op	      48 allocs/op
BenchmarkWrite/gzip-16         	     106	  10503831 ns/op	  467140 B/op	      98 allocs/op
BenchmarkWrite/gzip-1-16       	     231	   5146374 ns/op	  448971 B/op	      98 allocs/op
BenchmarkWrite/lz4-16          	     427	   2878722 ns/op	  269449 B/op	      58 allocs/op
BenchmarkWrite/snappy-16       	     715	   1721530 ns/op	  750768 B/op	      56 allocs/op
BenchmarkWrite/snappyv2-16     	     296	   5321221 ns/op	10866139 B/op	     158 allocs/op
PASS
ok  	github.com/grafana/loki/pkg/chunkenc	29.111s
```

### CODE REVIEW

File: pkg/chunkenc/memchunk.go [hunk 1]
```diff
@@ -140,34 +152,23 @@ func NewMemChunkSize(enc Encoding, blockSize, targetSize int) *MemChunk {
...
-// NewMemChunk returns a new in-mem chunk for query.
-func NewMemChunk(enc Encoding) *MemChunk {
-	return NewMemChunkSize(enc, 256*1024, 0)
-}
-
 // NewByteChunk returns a MemChunk on the passed bytes.
 func NewByteChunk(b []byte) (*MemChunk, error) {
 	bc := &MemChunk{
-		cPool:    &Gzip,
 		encoding: EncGZIP,
```
Comment (slim-bean) [inline pkg/chunkenc/memchunk.go]: I'm a little confused as to why this is set to EncGZIP here (maybe to default always to GZIP?) and I don't see anywhere that it gets reset if the encoding is different?

Comment (cyriltovena) [reply pkg/chunkenc/memchunk.go]: yes this is leftover.

Comment (cyriltovena) [reply pkg/chunkenc/memchunk.go]: Thanks !

File: pkg/chunkenc/util_test.go [hunk 2]
```diff
@@ -0,0 +1,1062 @@
...
+		i++
+		inserted += int64(len(entry.Line))
+		entry.Timestamp = time.Unix(0, i)
+		entry.Line = logString(i)
+
+	}
+	return inserted
+}
+
+var logs = strings.Split(`level=info ts=2019-12-12T15:00:08.325Z caller=compact.go:441 component=tsdb msg="compact blocks" count=3 mint=1576130400000 maxt=1576152000000 ulid=01DVX9ZHNM71GRCJS7M34Q0EV7 sources="[01DVWNC6NWY1A60AZV3Z6DGS65 01DVWW7XXX75GHA6ZDTD170CSZ 01DVX33N5W86CWJJVRPAVXJRWJ]" duration=2.897213221s
```
Comment (pstibrany) [inline pkg/chunkenc/util_test.go]: This belongs to testdata package.

File: pkg/ingester/ingester.go [hunk 3]
```diff
@@ -65,6 +67,7 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.DurationVar(&cfg.MaxChunkIdle, "ingester.chunks-idle-period", 30*time.Minute, "")
 	f.IntVar(&cfg.BlockSize, "ingester.chunks-block-size", 256*1024, "")
 	f.IntVar(&cfg.TargetChunkSize, "ingester.chunk-target-size", 0, "")
+	f.StringVar(&cfg.ChunkEncoding, "ingester.chunk-encoding", chunkenc.EncGZIP.String(), "The algorithm to use for compressing chunk. (gzip, lz4, snappy)")
```
Comment (owen-d) [inline pkg/ingester/ingester.go]: this should probably use the range of `supportedEncoding`s `Stringer` methods from `pkg/chunkenc` for the description.

**Comment by isavcic** [comment]
I see zstandard missing from this list, any particular reason?

**Comment by cyriltovena** [comment]
It was there, but the performance were bad so we removed it.

**Pull request merged**